### PR TITLE
llvm: add `BOLT` related tools

### PR DESCRIPTION
**Pull request by sabify**
This PR adds [`BOLT`](https://github.com/llvm/llvm-project/tree/main/bolt) related binaries like `llvm-bolt` and `merge-fdata` in `LLVM` toolchain.

### CODE REVIEW

File: Formula/llvm.rb [hunk 1]
```diff
@@ -59,6 +59,7 @@ def install
       lldb
       mlir
       polly
+      bolt
```
Comment (carlocab) [inline Formula/llvm.rb:62]: 1. Doesn't this need to be restricted to Linux? Last I checked, Bolt supports only ELF files.
2. Does this install properly? It seems the install configuration is still in flux upstream: https://github.com/llvm/llvm-project/commit/90dcdc4b6e7d86cb3d5049bd766aecddd549dd7d, https://github.com/llvm/llvm-project/commit/ad8c34bc3089d847a09bb740f7a58c96073e0959, https://github.com/llvm/llvm-project/commit/f7a33090a91015836497c75f173775392ab0304d, https://github.com/llvm/llvm-project/commit/07b749800c5cd4105d49ab46be5f0a2079dd709a, https://github.com/llvm/llvm-project/commit/0bb317b7bff3d4f70bddd03a1f337c3f2a118943
3. Is this included in the builds that upstream ship? (e.g. the builds that can be downloaded [here](https://github.com/llvm/llvm-project/releases/tag/llvmorg-14.0.6), or in the [APT packages](https://apt.llvm.org))
4. Can we add a test?

Also, LLVM takes ages to build, so we'll probably just include this in #106925, if it is suitable for packaging.

**Review by carlocab — COMMENTED** [review]
Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

**Comment by sabify** [comment]
I close this PR. Let's add this feature on your mentioned PR, LLVM 15.

**Comment by carlocab** [comment]
> I close this PR. Let's add this feature on your mentioned PR, LLVM 15.

I'd still need answers to my questions here: https://github.com/Homebrew/homebrew-core/pull/108795#discussion_r953966081

In particular, I'd need you to supply a test.

File: Formula/llvm.rb (hunk 1 above)
Comment (sabify) [reply Formula/llvm.rb:62]: 1. Yes, it only supports ELF binaries and I should've added it under Linux OS condition.
2. With Homebrew build system, I'm not sure. I have to try it.
3. No, it is not included in the official LLVM package (I've tried the latest APT package, LLVM 15).
4. You can but I don't think the optimized binary generated by `llvm-bolt` be consistent on new releases. So, tracking the test results would be bothersome.

Comment (carlocab) [reply Formula/llvm.rb:62]: > 4. You can but I don't think the optimized binary generated by `llvm-bolt` be consistent on new releases. So, tracking the test results would be bothersome.

We don't need to track this that closely. We should at least check that:
1. It installed _something_
2. That something can be used and doesn't crash when you try to make basic use of it
If doing the above is still bothersome then it's definitely not a good idea to install this.

Comment (sabify) [reply Formula/llvm.rb:62]: Oh! in those cases you've mentioned, it's simple to write a test for it.

I'll try to propose changes in your llvm-15 PR.

**Comment by sabify** [comment]
@carlocab I'm trying to build the `llvm` with `-DLLVM_ENABLE_ASSERTIONS=ON` flag which is needed by `bolt`, but I get bunch of link errors. Could you please check [the cmake error log](https://www.mediafire.com/file/qvy4scm5a89alhb/CMakeError.log/file) and let me know what's happening there? I see -lgcc not found there too!

**Comment by carlocab** [comment]
If you're getting link errors then the problem won't be in `CMakeError.log` -- AFAIU that's for errors during configure-time.

In any case,

> build the `llvm` with `-DLLVM_ENABLE_ASSERTIONS=ON` flag which is needed by `bolt`

This is probably a deal-breaker. We don't want to enable assertions in our LLVM build.

**Comment by sabify** [comment]
That makes sense. How about providing `llvm-bolt` as a new formula? Its versioning would be the same as `llvm`.

**Comment by carlocab** [comment]
That sounds like it would need its own build of LLVM with `-DLLVM_ENABLE_ASSERTIONS=ON`, which we don't want to do. There are already several different builds of LLVM in distinct formulae in this repository. We shouldn't add another one.

**Pull request closed without merging**